File: algorithms/medium/last_stone_weight_ii.ex
defmodule Solution do
  @spec last_stone_weight_ii(stones :: [integer]) :: integer
  def last_stone_weight_ii(stones) do
    sum = Enum.sum(stones)
    target = div(sum, 2)

    for stone <- stones, j <- target..stone//-1, reduce: %{} do
      dp -> Map.put(dp, j, max(Map.get(dp, j, 0), Map.get(dp, j - stone, 0) + stone))
    end
    |> Map.get(target, 0)
    |> then(&(sum - 2 * &1))
  end
end

File: algorithms/medium/last_stone_weight_ii.rs
impl Solution {
    pub fn last_stone_weight_ii(stones: Vec<i32>) -> i32 {
        let sum: i32 = stones.iter().sum();
        let target = (sum / 2) as usize;
        let mut dp = vec![0; target + 1];
        for stone in stones {
            for j in (stone as usize..=target).rev() {
                dp[j] = dp[j].max(dp[j-stone as usize] + stone);
            }
        }

        sum - 2 * dp[target]
    }
}
